code/obj/item/zoldorfmisc.dm

```diff
@@ -15,18 +34,34 @@
 			src.desc = "This is one WEIRD burrito..."
 
 	attackby(obj/item/weapon, mob/user)
-		if(istype(weapon, /obj/item/pen) && src.icon_state=="scrollopen")
-			user.visible_message(SPAN_ALERT("<b>[user.name] stabs themself with the [weapon] and starts signing the contract in blood!</b>"),SPAN_ALERT("<b>You stab yourself with the [weapon] and start signing the contract in blood!</b>"))
-			playsound(user, 'sound/impact_sounds/Flesh_Stab_1.ogg', 60, TRUE)
-			take_bleeding_damage(user, null, 10, DAMAGE_STAB)
-			src.icon_state = "signing"
-			if (do_after(user, 4.6 SECONDS))
-				user.visible_message(SPAN_ALERT("<b>[user.name] finishes signing the contract in blood!</b>"),SPAN_ALERT("<b>You finish signing the contract in blood!</b>"))
-				src.signer = user.real_name
-				src.name = "[user.real_name]'s signed demonic contract"
-				src.icon_state = "signed"
-			else
-				src.icon_state = "scrollopen"
+		if (!istype(weapon, /obj/item/pen) || src.signing_state != UNSIGNED)
+			return
+		user.visible_message(
+			SPAN_ALERT("<b>[user.name] stabs [himself_or_herself(user)] with the [weapon] and starts signing the contract in blood!</b>"),
+			SPAN_ALERT("<b>You stab yourself with the [weapon] and start signing the contract in blood!</b>"))
+		src.signing_state = SIGNING
+		playsound(user, 'sound/impact_sounds/Flesh_Stab_1.ogg', 60, TRUE)
+		take_bleeding_damage(user, null, 10, DAMAGE_STAB)
+		src.icon_state = "signing"
+		var/actionBar = new /datum/action/bar/icon/callback/signingBar(
+			user,
+			src,
+			SIGNING_DURATION,
+			PROC_REF(complete_signing),
+			list(user),
+			src.icon,
+			src.icon_state,
+			"<b>[user.name] finishes signing the contract in blood!</b>",
+			(INTERRUPT_ACT | INTERRUPT_STUNNED | INTERRUPT_ACTION | INTERRUPT_MOVE))
```

Comment:
any reason not to use the `SETUP_GENERIC_ACTIONBAR` macro?

Reply:
I would be happy to except the old code changed the icon state if the action was interrupted `src.icon_state = "scrollopen"`. Wasn't able to see a way to do that with SETUP_GENERIC_ACTIONBAR.

Reply:
mm. In that case, I think it would be better to make the action a direct child of `/datum/action/bar/icon` rather than `...callback`. Mostly a nitpick but it would be a bit cleaner

Reply:
Done.